typescript/sdk/src/core/HyperlaneCore.ts

```diff
@@ -102,31 +112,28 @@ export class HyperlaneCore extends HyperlaneApp<CoreFactories> {
     return ismReader.deriveIsmConfig(address);
   }
 
-  async buildMetadata(message: DispatchedMessage): Promise<string> {
+  async getHookConfig(message: DispatchedMessage): Promise<DerivedHookConfig> {
+    const originChain = this.getOrigin(message);
+    const hookReader = new EvmHookReader(this.multiProvider, originChain);
+    const address = await this.getHookAddress(message);
+    return hookReader.deriveHookConfig(address);
+  }
+
+  async buildMetadata(
+    message: DispatchedMessage,
+    dispatchTx: TransactionReceipt,
+  ): Promise<string> {
     const ismConfig = await this.getRecipientIsmConfig(message);
-    const destinationChain = this.getDestination(message);
+    const hookConfig = await this.getHookConfig(message);
 
-    switch (ismConfig.type) {
-      case IsmType.TRUSTED_RELAYER:
-        // eslint-disable-next-line no-case-declarations
-        const destinationSigner = await this.multiProvider.getSignerAddress(
-          destinationChain,
-        );
-        if (!eqAddress(destinationSigner, ismConfig.relayer)) {
-          this.logger.warn(
-            `${destinationChain} signer ${destinationSigner} does not match trusted relayer ${ismConfig.relayer}`,
-          );
-        }
-    }
+    const baseMetadataBuilder = new BaseMetadataBuilder(this);
 
-    // TODO: implement metadata builders for other module types
-    const moduleType = ismTypeToModuleType(ismConfig.type);
-    switch (moduleType) {
-      case ModuleType.NULL:
-        return '0x';
-      default:
-        throw new Error(`Unsupported module type ${moduleType}`);
-    }
+    return baseMetadataBuilder.build({
```

Comment:
note that we currently have the tests skipped for `baseMetadataBuilder.build`, it mostly works but we disabled because the randomness can catch errors sometimes - https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/3941

Reply:
kek right on queue https://github.com/hyperlane-xyz/hyperlane-monorepo/actions/runs/9502728901/job/26191661756?pr=3954#step:7:35906

might wanna cherry-pick https://github.com/hyperlane-xyz/hyperlane-monorepo/commit/29f368513faa73f041be4b00906c086c70384fa5

Reply:
thanks!